{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Loading / clustering data functions\n",
    "import pandas as pd\n",
    "import Bio.Phylo as bp\n",
    "import scipy as sc\n",
    "\n",
    "class TrueTree:\n",
    "    \"\"\"\n",
    "    Data-like class meant to load the True Tree from the .phy file accompanying this package.\n",
    "    \n",
    "    Attributes:\n",
    "        true_tree (Bio.Phylo.Tree): The true tree stored in a Bio.Phylo tree class.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        \"\"\"\n",
    "        Constructor for TrueTree data class.\n",
    "        \n",
    "        Paramters:\n",
    "            None\n",
    "        \"\"\"\n",
    "        self.true_tree = []\n",
    "        \n",
    "    def load_true_tree(self,file_path):\n",
    "        \"\"\"\n",
    "        Function to load and parse the true tree, and store it.\n",
    "        \n",
    "        Parameters:\n",
    "            file_path (string): A valid file path to the true tree .phy file\n",
    "         \n",
    "        Returns:\n",
    "            Bio.Phylo.Tree: A Bio.Phylo Tree class storing the true tree.\n",
    "        \"\"\"\n",
    "        self.true_tree = bp.read(file_path,'newick')\n",
    "        return(self.true_tree)\n",
    "        \n",
    "class CNVData:\n",
    "    \"\"\"\n",
    "    Data-like class meant to store a Species by Gene matrix with entries corresponding to gene Copy Number. A specific datafile is provided for working with reptilian species samples.\n",
    "    \n",
    "    Attributes:\n",
    "        CNVData (pandas.DataFrame): A pandas DataFrame containing the CNV data matrix, with species as rows, and genes as columns. For this specific data-set, there are 91 species and 6491 orthologous genes across the 91 species. \n",
    "        \n",
    "        label_replace_dictionary (dict): A dictionary storing phylogenetically correct names for some species in the given CNV data matrix. \n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        \"\"\"\n",
    "        Constructor for CNVData data class.\n",
    "        \n",
    "        Paramters:\n",
    "            None\n",
    "        \"\"\"\n",
    "        \n",
    "        self.CNVData = []\n",
    "        self.label_replace_dictionary = {\n",
    "        'SM1_centroids_nucleotides': 'Chelydra_serpentina',\n",
    "        'SM2_centroids_nucleotides': 'Anolis_sagrei',\n",
    "        'SM3_centroids_nucleotides': 'Elgaria_multicarinata',\n",
    "        'SM4_centroids_nucleotides': 'Lamprophis',\n",
    "        'SM5_centroids_nucleotides': 'Agkistrodon_piscivorus',\n",
    "        'SM6_centroids_nucleotides': 'Xenopeltis_unicolor',\n",
    "        'SM7_centroids_nucleotides': 'Alligator_mississippiensis',\n",
    "        'SM8_centroids_nucleotides': 'Sceloporus_undulatus',\n",
    "        'SM9_centroids_nucleotides': 'Pogona',\n",
    "        'SM10_centroids_nucleotides': 'Sternotherus_odoratus',\n",
    "        'SM11_centroids_nucleotides': 'Sternotherus_odoratus',\n",
    "        'SM12_centroids_nucleotides': 'Scincella_lateralis',\n",
    "        'SM13_centroids_nucleotides': 'Terrapene_carolina',\n",
    "        'SM14_centroids_nucleotides': 'Agkistrodon_piscivorus',\n",
    "        'SM15_centroids_nucleotides': 'Eublepharis_macularius',\n",
    "        'TC_centroids_nucleotides': 'Thamnophis_sirtalis',\n",
    "        'HS08_centroids_nucleotides': 'Thamnophis_elegans',\n",
    "        'HS11_centroids_nucleotides': 'Thamnophis_couchii'\n",
    "        }\n",
    "\n",
    "    \n",
    "    def readCNVMatrix(self,file_path):\n",
    "        \"\"\"\n",
    "        Function to read the CNV data matrix from a .tsv file, and parse it correctly. NA's are replaced with 0s, and the copy numbers are min-max normalized across their species as a standardization method. Incorrectly named species are renamed using self.relabelIndex.\n",
    "        \n",
    "        Parameters:\n",
    "            file_path (string): A valid file path to the .tsv CNV data file\n",
    "            \n",
    "        Returns:\n",
    "            self.CNVData (pandas.DataFrame):  A pandas DataFrame containing the CNV data matrix, with species as rows, and genes as columns. For this specific data-set, there are 91 species and 6491 orthologous genes across the 91 species. The copy number values are normalized across species is heavily zero-inflated.\n",
    "            \n",
    "        \"\"\"\n",
    "        ## Read in data-file using Pandas. Assume TSV. Use column 1 as index. \n",
    "        ## Transpose so genes are columns. \n",
    "        self.CNVData = pd.read_csv(file_path,sep='\\t', index_col=0).transpose()\n",
    "        \n",
    "        ## Drop unneccesary Column counting total CNV per species\n",
    "        self.CNVData = self.CNVData.drop(['Total'])\n",
    "        ## Replace NaN entries --> 0\n",
    "        self.CNVData.fillna(0,inplace=True) \n",
    "        \n",
    "        ## Min/max scaling of copy number\n",
    "        self.CNVData = (self.CNVData - self.CNVData.min()) / (self.CNVData.max()-self.CNVData.min()) ## Min-max transform\n",
    "        \n",
    "        ## Replace NaN entries --> 0\n",
    "        self.CNVData.fillna(0,inplace=True) \n",
    "        \n",
    "        ## Relabel species names to line up with 'true' tree\n",
    "        self.CNVData.index = self.relabelIndex(self.CNVData.index)\n",
    "        \n",
    "        return(self.CNVData)\n",
    "    \n",
    "    def relabelIndex(self,labels):\n",
    "        \"\"\"\n",
    "        Function to relabel the indicies of a pandas DataFrame containing reptile species names with their correct phylogenetic names using the class attribute self.label_replace_dictionary.\n",
    "        \n",
    "        Parameters:\n",
    "            labels (pandas.DataFrame.index): An index of a pandas DataFrame containing reptile species names. Behaves like a list object \n",
    "            \n",
    "        Returns:\n",
    "            fixed_labels (list): A list object containing the corrected phylogenetic names of reptile species passed in labels.\n",
    "        \n",
    "        \"\"\"\n",
    "        ## init correct name matrix\n",
    "        fixed_labels = []\n",
    "\n",
    "        ## Iterate over the labels\n",
    "        for label in labels:\n",
    "            ## if it must be fixed, lookup in label_replace_dictionary and store the fixed name \n",
    "            ## or store the name if not.\n",
    "            if label in self.label_replace_dictionary:\n",
    "                fixed_labels.append(self.label_replace_dictionary[label])\n",
    "            else:\n",
    "                fixed_labels.append(label)\n",
    "        return(fixed_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.cluster as sc\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "\n",
    "class Cluster: \n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    def __init__(self): ## takes a CNVData matrix. \n",
    "        self.trees = []\n",
    "        self.clust_ids=[]\n",
    "        self.random_trees=[]\n",
    "        self.random_clust_ids=[]\n",
    "        self.available_methods = ['single','complete','average',\n",
    "                                  'weighted','ward','median','centroid']\n",
    "        \n",
    "    def cluster(self,data,method=None):\n",
    "        \"\"\"   \n",
    "        Function to use \n",
    "\n",
    "        \"\"\"\n",
    "        ## try to load the method. \n",
    "        if method:\n",
    "            try:\n",
    "                self.available_methods.index(method)\n",
    "            except ValueError:\n",
    "                raise Exception('Selected clustering method not available.')\n",
    "        else:\n",
    "            method = 'ward'\n",
    "        \n",
    "        \n",
    "        rand = data.copy()\n",
    "        num_zeroes = (data.values==0).sum()\n",
    "        val_max = data.values.max()\n",
    " \n",
    "        for x in range(0,100):\n",
    "            \n",
    "            boot_tree, boot_clust_id = self.cluster_single(data.sample(data.shape[1],replace=True,axis=1),method)\n",
    "            \n",
    "            self.trees.append(boot_tree)\n",
    "            self.clust_ids.append(boot_clust_id)\n",
    "            \n",
    "            rand_tree, rand_clust_id = self.cluster_single(self.generate_random(rand,val_max,num_zeroes),method)\n",
    "            \n",
    "            self.random_trees.append(rand_tree)\n",
    "            self.random_clust_ids.append(rand_clust_id)\n",
    "                    \n",
    "        return self.trees, self.random_trees, self.clust_ids, self.random_clust_ids\n",
    "\n",
    "    def cluster_single(self,data,method):\n",
    "        link = sc.hierarchy.linkage(data,method)\n",
    "        tree = sc.hierarchy.to_tree(link)\n",
    "        \n",
    "        cluster_id = dict(zip(data.index,sc.hierarchy.cut_tree(link,n_clusters=4).T.tolist()[0]))\n",
    "        \n",
    "        cluster_newick = self.getNewick(tree, \"\", tree.dist, data.index)\n",
    "        cluster_phylo = bp.read(StringIO(cluster_newick),'newick')\n",
    "        return cluster_phylo, cluster_id\n",
    "\n",
    "    \n",
    "    def getNewick(self,node, newick, parentdist, leaf_names):\n",
    "        \"\"\"\n",
    "        Function to \n",
    "        \"\"\"\n",
    "        if node.is_leaf():\n",
    "            return \"%s:%.2f%s\" % (leaf_names[node.id], parentdist - node.dist, newick)\n",
    "        else:\n",
    "            if len(newick) > 0:\n",
    "                newick = \"):%.2f%s\" % (parentdist - node.dist, newick)\n",
    "            else:\n",
    "                newick = \");\"\n",
    "            newick = self.getNewick(node.get_left(), newick, node.dist, leaf_names)\n",
    "            newick = self.getNewick(node.get_right(), \",%s\" % (newick), node.dist, leaf_names)\n",
    "            newick = \"(%s\" % (newick)\n",
    "            return newick\n",
    "        \n",
    "    def generate_random(self,rm,maxval,zeroes):\n",
    "        randarray = np.random.random_sample(rm.values.size)*maxval\n",
    "        randarray[np.random.choice(np.arange(0,rm.values.size-1),zeroes,replace=False)] = 0\n",
    "        rm.iloc[:,:] = randarray.reshape(rm.shape)\n",
    "        return(rm)\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = TrueTree().load_true_tree('data/phyliptree.phy')\n",
    "dm = CNVData().readCNVMatrix('data/LS_blastn_Gar_noDenom.txt')\n",
    "cl,rn,cl_id,rn_id = Cluster().cluster(dm,'ward')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.hierarchy.linkage()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
