{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Loading / clustering data functions\n",
    "import pandas as pd\n",
    "import Bio.Phylo as bp\n",
    "import scipy as sc\n",
    "\n",
    "class TrueTree:\n",
    "    \"\"\"\n",
    "    Data-like class meant to load the True Tree from the .phy file accompanying this package.\n",
    "    \n",
    "    Attributes:\n",
    "        true_tree (Bio.Phylo.Tree): The true tree stored in a Bio.Phylo tree class.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self):\n",
    "        \"\"\"\n",
    "        Constructor for TrueTree data class.\n",
    "        \n",
    "        Paramters:\n",
    "            None\n",
    "        \"\"\"\n",
    "        self.true_tree = []\n",
    "        \n",
    "    def load_true_tree(self,file_path):\n",
    "        \"\"\"\n",
    "        Function to load and parse the true tree, and store it.\n",
    "        \n",
    "        Parameters:\n",
    "            file_path (string): A valid file path to the true tree .phy file\n",
    "         \n",
    "        Returns:\n",
    "            Bio.Phylo.Tree: A Bio.Phylo Tree class storing the true tree.\n",
    "        \"\"\"\n",
    "        self.true_tree = bp.read(file_path,'newick')\n",
    "        return(self.true_tree)\n",
    "        \n",
    "class CNVData:\n",
    "    \"\"\"\n",
    "    Data-like class meant to store a Species by Gene matrix with entries corresponding to gene Copy Number. A specific datafile is provided for working with reptilian species samples.\n",
    "    \n",
    "    Attributes:\n",
    "        CNVData (pandas.DataFrame): A pandas DataFrame containing the CNV data matrix, with species as rows, and genes as columns. For this specific data-set, there are 91 species and 6491 orthologous genes across the 91 species. \n",
    "        \n",
    "        label_replace_dictionary (dict): A dictionary storing phylogenetically correct names for some species in the given CNV data matrix. \n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        \"\"\"\n",
    "        Constructor for CNVData data class.\n",
    "        \n",
    "        Paramters:\n",
    "            None\n",
    "        \"\"\"\n",
    "        \n",
    "        self.CNVData = []\n",
    "        self.label_replace_dictionary = {\n",
    "        'SM1_centroids_nucleotides': 'Chelydra_serpentina',\n",
    "        'SM2_centroids_nucleotides': 'Anolis_sagrei',\n",
    "        'SM3_centroids_nucleotides': 'Elgaria_multicarinata',\n",
    "        'SM4_centroids_nucleotides': 'Lamprophis',\n",
    "        'SM5_centroids_nucleotides': 'Agkistrodon_piscivorus',\n",
    "        'SM6_centroids_nucleotides': 'Xenopeltis_unicolor',\n",
    "        'SM7_centroids_nucleotides': 'Alligator_mississippiensis',\n",
    "        'SM8_centroids_nucleotides': 'Sceloporus_undulatus',\n",
    "        'SM9_centroids_nucleotides': 'Pogona',\n",
    "        'SM10_centroids_nucleotides': 'Sternotherus_odoratus',\n",
    "        'SM11_centroids_nucleotides': 'Sternotherus_odoratus',\n",
    "        'SM12_centroids_nucleotides': 'Scincella_lateralis',\n",
    "        'SM13_centroids_nucleotides': 'Terrapene_carolina',\n",
    "        'SM14_centroids_nucleotides': 'Agkistrodon_piscivorus',\n",
    "        'SM15_centroids_nucleotides': 'Eublepharis_macularius',\n",
    "        'TC_centroids_nucleotides': 'Thamnophis_sirtalis',\n",
    "        'HS08_centroids_nucleotides': 'Thamnophis_elegans',\n",
    "        'HS11_centroids_nucleotides': 'Thamnophis_couchii'\n",
    "        }\n",
    "\n",
    "    \n",
    "    def readCNVMatrix(self,file_path):\n",
    "        \"\"\"\n",
    "        Function to read the CNV data matrix from a .tsv file, and parse it correctly. NA's are replaced with 0s, and the copy numbers are min-max normalized across their species as a standardization method. Incorrectly named species are renamed using self.relabelIndex.\n",
    "        \n",
    "        Parameters:\n",
    "            file_path (string): A valid file path to the .tsv CNV data file\n",
    "            \n",
    "        Returns:\n",
    "            self.CNVData (pandas.DataFrame):  A pandas DataFrame containing the CNV data matrix, with species as rows, and genes as columns. For this specific data-set, there are 91 species and 6491 orthologous genes across the 91 species. The copy number values are normalized across species is heavily zero-inflated.\n",
    "            \n",
    "        \"\"\"\n",
    "        ## Read in data-file using Pandas. Assume TSV. Use column 1 as index. \n",
    "        ## Transpose so genes are columns. \n",
    "        self.CNVData = pd.read_csv(file_path,sep='\\t', index_col=0).transpose()\n",
    "        \n",
    "        ## Drop unneccesary Column counting total CNV per species\n",
    "        self.CNVData = self.CNVData.drop(['Total'])\n",
    "        ## Replace NaN entries --> 0\n",
    "        self.CNVData.fillna(0,inplace=True) \n",
    "        \n",
    "        ## Min/max scaling of copy number\n",
    "        self.CNVData = (self.CNVData - self.CNVData.min()) / (self.CNVData.max()-self.CNVData.min()) ## Min-max transform\n",
    "        \n",
    "        ## Replace NaN entries --> 0\n",
    "        self.CNVData.fillna(0,inplace=True) \n",
    "        \n",
    "        ## Relabel species names to line up with 'true' tree\n",
    "        self.CNVData.index = self.relabelIndex(self.CNVData.index)\n",
    "        \n",
    "        return(self.CNVData)\n",
    "    \n",
    "    def relabelIndex(self,labels):\n",
    "        \"\"\"\n",
    "        Function to relabel the indicies of a pandas DataFrame containing reptile species names with their correct phylogenetic names using the class attribute self.label_replace_dictionary.\n",
    "        \n",
    "        Parameters:\n",
    "            labels (pandas.DataFrame.index): An index of a pandas DataFrame containing reptile species names. Behaves like a list object \n",
    "            \n",
    "        Returns:\n",
    "            fixed_labels (list): A list object containing the corrected phylogenetic names of reptile species passed in labels.\n",
    "        \n",
    "        \"\"\"\n",
    "        ## init correct name matrix\n",
    "        fixed_labels = []\n",
    "\n",
    "        ## Iterate over the labels\n",
    "        for label in labels:\n",
    "            ## if it must be fixed, lookup in label_replace_dictionary and store the fixed name \n",
    "            ## or store the name if not.\n",
    "            if label in self.label_replace_dictionary:\n",
    "                fixed_labels.append(self.label_replace_dictionary[label])\n",
    "            else:\n",
    "                fixed_labels.append(label)\n",
    "        return(fixed_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.cluster as sc\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "\n",
    "class Cluster: \n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    def __init__(self): ## takes a CNVData matrix. \n",
    "        self.trees = []\n",
    "        self.clust_ids=[]\n",
    "        self.random_trees=[]\n",
    "        self.random_clust_ids=[]\n",
    "        self.available_methods = ['single','complete','average',\n",
    "                                  'weighted','ward','median','centroid']\n",
    "        self.available_distances=['braycurtis', 'canberra', 'chebyshev', \n",
    "                                  'cityblock', 'correlation', 'cosine', \n",
    "                                  'dice', 'euclidean', 'hamming', 'jaccard', \n",
    "                                  'jensenshannon', 'kulsinski', 'mahalanobis',\n",
    "                                  'matching', 'minkowski', 'rogerstanimoto', \n",
    "                                  'russellrao', 'seuclidean', 'sokalmichener', \n",
    "                                  'sokalsneath', 'sqeuclidean', 'yule']\n",
    "        \n",
    "    def cluster(self,data,method=None,distance=None):\n",
    "        \"\"\"   \n",
    "        Function to use \n",
    "\n",
    "        \"\"\"\n",
    "        ## try to load the method. \n",
    "        if method:\n",
    "            try:\n",
    "                self.available_methods.index(method)\n",
    "            except ValueError:\n",
    "                raise Exception('Selected clustering method not available.')\n",
    "        else:\n",
    "            method = 'ward'\n",
    "          \n",
    "        if distance:\n",
    "            try:\n",
    "                self.available_distances.index(distance)\n",
    "            except ValueError:\n",
    "                raise Exception('Selected distance metric not available.')\n",
    "        else:\n",
    "            distance = 'euclidean'\n",
    "            \n",
    "        \n",
    "        rand = data.copy()\n",
    "        num_zeroes = (data.values==0).sum()\n",
    "        val_max = data.values.max()\n",
    " \n",
    "        for x in range(0,100):\n",
    "            \n",
    "            boot_tree, boot_clust_id = self.cluster_single(data.sample(data.shape[1],replace=True,axis=1),method)\n",
    "            \n",
    "            self.trees.append(boot_tree)\n",
    "            self.clust_ids.append(boot_clust_id)\n",
    "            \n",
    "            rand_tree, rand_clust_id = self.cluster_single(self.generate_random(rand,val_max,num_zeroes),method)\n",
    "            \n",
    "            self.random_trees.append(rand_tree)\n",
    "            self.random_clust_ids.append(rand_clust_id)\n",
    "                    \n",
    "        return self.trees, self.random_trees, self.clust_ids, self.random_clust_ids\n",
    "\n",
    "    def cluster_single(self,data,method):\n",
    "        link = sc.hierarchy.linkage(data,method)\n",
    "        tree = sc.hierarchy.to_tree(link)\n",
    "        \n",
    "        cluster_id = dict(zip(data.index,sc.hierarchy.cut_tree(link,n_clusters=4).T.tolist()[0]))\n",
    "        \n",
    "        cluster_newick = self.getNewick(tree, \"\", tree.dist, data.index)\n",
    "        cluster_phylo = bp.read(StringIO(cluster_newick),'newick')\n",
    "        return cluster_phylo, cluster_id\n",
    "\n",
    "    \n",
    "    def getNewick(self,node, newick, parentdist, leaf_names):\n",
    "        \"\"\"\n",
    "        Function to \n",
    "        \"\"\"\n",
    "        if node.is_leaf():\n",
    "            return \"%s:%.2f%s\" % (leaf_names[node.id], parentdist - node.dist, newick)\n",
    "        else:\n",
    "            if len(newick) > 0:\n",
    "                newick = \"):%.2f%s\" % (parentdist - node.dist, newick)\n",
    "            else:\n",
    "                newick = \");\"\n",
    "            newick = self.getNewick(node.get_left(), newick, node.dist, leaf_names)\n",
    "            newick = self.getNewick(node.get_right(), \",%s\" % (newick), node.dist, leaf_names)\n",
    "            newick = \"(%s\" % (newick)\n",
    "            return newick\n",
    "        \n",
    "    def generate_random(self,rm,maxval,zeroes):\n",
    "        randarray = np.random.random_sample(rm.values.size)*maxval\n",
    "        randarray[np.random.choice(np.arange(0,rm.values.size-1),zeroes,replace=False)] = 0\n",
    "        rm.iloc[:,:] = randarray.reshape(rm.shape)\n",
    "        return(rm)\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "tr = TrueTree().load_true_tree('data/phyliptree.phy')\n",
    "dm = CNVData().readCNVMatrix('data/LS_blastn_Gar_noDenom.txt')\n",
    "cl,rn,cl_id,rn_id = Cluster().cluster(dm,'ward')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  9.        ,  16.        ,  19.32021106,   2.        ],\n",
       "       [ 60.        ,  77.        ,  21.79133365,   2.        ],\n",
       "       [ 23.        ,  87.        ,  21.8410419 ,   2.        ],\n",
       "       [ 84.        ,  93.        ,  21.88572594,   3.        ],\n",
       "       [ 59.        ,  85.        ,  22.15481513,   2.        ],\n",
       "       [ 19.        ,  94.        ,  22.48319125,   4.        ],\n",
       "       [ 33.        ,  76.        ,  22.6302342 ,   2.        ],\n",
       "       [ 68.        ,  92.        ,  22.81990481,   3.        ],\n",
       "       [ 25.        ,  91.        ,  22.85683098,   3.        ],\n",
       "       [ 48.        ,  51.        ,  22.94515805,   2.        ],\n",
       "       [ 20.        ,  97.        ,  23.11894365,   3.        ],\n",
       "       [ 50.        ,  89.        ,  23.31820938,   2.        ],\n",
       "       [ 10.        ,  57.        ,  23.32099078,   2.        ],\n",
       "       [ 55.        ,  63.        ,  23.42671196,   2.        ],\n",
       "       [ 32.        ,  52.        ,  23.59099404,   2.        ],\n",
       "       [ 15.        , 104.        ,  23.61718419,   3.        ],\n",
       "       [102.        , 103.        ,  23.63489652,   4.        ],\n",
       "       [ 95.        , 101.        ,  23.71757436,   5.        ],\n",
       "       [ 75.        ,  96.        ,  23.71795499,   5.        ],\n",
       "       [ 12.        ,  73.        ,  23.7758339 ,   2.        ],\n",
       "       [ 13.        ,  99.        ,  23.78172735,   4.        ],\n",
       "       [ 29.        , 105.        ,  23.7953135 ,   3.        ],\n",
       "       [ 18.        , 109.        ,  23.80242751,   6.        ],\n",
       "       [ 37.        , 106.        ,  23.82407746,   4.        ],\n",
       "       [107.        , 108.        ,  23.82784902,   9.        ],\n",
       "       [ 28.        , 115.        ,  23.83845516,  10.        ],\n",
       "       [  1.        , 116.        ,  23.86462398,  11.        ],\n",
       "       [ 44.        ,  61.        ,  23.87054787,   2.        ],\n",
       "       [ 53.        , 100.        ,  23.88816536,   3.        ],\n",
       "       [113.        , 117.        ,  23.90265559,  17.        ],\n",
       "       [ 34.        ,  83.        ,  24.01228274,   2.        ],\n",
       "       [ 78.        , 114.        ,  24.09355608,   5.        ],\n",
       "       [112.        , 122.        ,  24.10274881,   8.        ],\n",
       "       [111.        , 119.        ,  24.13339664,   7.        ],\n",
       "       [120.        , 123.        ,  24.14439203,  25.        ],\n",
       "       [ 86.        , 125.        ,  24.17733098,  26.        ],\n",
       "       [  4.        , 126.        ,  24.17963445,  27.        ],\n",
       "       [ 47.        , 127.        ,  24.23402108,  28.        ],\n",
       "       [110.        , 128.        ,  24.23709852,  30.        ],\n",
       "       [124.        , 129.        ,  24.26971707,  37.        ],\n",
       "       [ 54.        , 130.        ,  24.27472394,  38.        ],\n",
       "       [ 17.        , 131.        ,  24.37489031,  39.        ],\n",
       "       [  2.        , 132.        ,  24.3823422 ,  40.        ],\n",
       "       [ 90.        , 121.        ,  24.51308041,   3.        ],\n",
       "       [ 65.        , 133.        ,  24.53846187,  41.        ],\n",
       "       [134.        , 135.        ,  24.60312198,  44.        ],\n",
       "       [ 74.        , 136.        ,  24.60641289,  45.        ],\n",
       "       [118.        , 137.        ,  24.6567773 ,  47.        ],\n",
       "       [ 26.        , 138.        ,  24.67092306,  48.        ],\n",
       "       [ 58.        , 139.        ,  24.70168685,  49.        ],\n",
       "       [ 14.        , 140.        ,  24.71918958,  50.        ],\n",
       "       [ 66.        ,  98.        ,  24.77184782,   4.        ],\n",
       "       [ 35.        ,  42.        ,  24.79816862,   2.        ],\n",
       "       [ 79.        , 141.        ,  24.80138885,  51.        ],\n",
       "       [142.        , 144.        ,  24.92745029,  55.        ],\n",
       "       [ 56.        , 145.        ,  24.96195995,  56.        ],\n",
       "       [143.        , 146.        ,  24.96444694,  58.        ],\n",
       "       [ 40.        , 147.        ,  25.01213039,  59.        ],\n",
       "       [ 27.        , 148.        ,  25.06732047,  60.        ],\n",
       "       [  5.        , 149.        ,  25.21202865,  61.        ],\n",
       "       [ 21.        , 150.        ,  25.24526358,  62.        ],\n",
       "       [ 39.        , 151.        ,  25.24963696,  63.        ],\n",
       "       [ 11.        , 152.        ,  25.28060302,  64.        ],\n",
       "       [ 88.        , 153.        ,  25.30919354,  65.        ],\n",
       "       [ 31.        , 154.        ,  25.31475196,  66.        ],\n",
       "       [ 81.        , 155.        ,  25.35319397,  67.        ],\n",
       "       [ 38.        ,  43.        ,  25.40902617,   2.        ],\n",
       "       [ 80.        , 156.        ,  25.46438144,  68.        ],\n",
       "       [  3.        , 158.        ,  25.50467277,  69.        ],\n",
       "       [ 46.        , 159.        ,  25.50982382,  70.        ],\n",
       "       [ 49.        , 160.        ,  25.6462148 ,  71.        ],\n",
       "       [ 30.        ,  41.        ,  25.70989866,   2.        ],\n",
       "       [161.        , 162.        ,  25.8075086 ,  73.        ],\n",
       "       [ 82.        , 163.        ,  25.85934175,  74.        ],\n",
       "       [ 24.        , 164.        ,  25.90423067,  75.        ],\n",
       "       [157.        , 165.        ,  25.92250738,  77.        ],\n",
       "       [ 36.        , 166.        ,  26.21171282,  78.        ],\n",
       "       [ 70.        ,  72.        ,  26.33316983,   2.        ],\n",
       "       [ 67.        , 168.        ,  26.4584895 ,   3.        ],\n",
       "       [  6.        , 167.        ,  26.47886998,  79.        ],\n",
       "       [  0.        , 170.        ,  26.54684016,  80.        ],\n",
       "       [ 69.        , 171.        ,  26.54927808,  81.        ],\n",
       "       [169.        , 172.        ,  26.73427367,  84.        ],\n",
       "       [ 22.        , 173.        ,  26.92944485,  85.        ],\n",
       "       [ 45.        , 174.        ,  27.13012368,  86.        ],\n",
       "       [  7.        , 175.        ,  27.32628466,  87.        ],\n",
       "       [ 71.        , 176.        ,  27.40301078,  88.        ],\n",
       "       [ 62.        , 177.        ,  27.49824742,  89.        ],\n",
       "       [ 64.        , 178.        ,  27.50986692,  90.        ],\n",
       "       [  8.        , 179.        ,  28.39998533,  91.        ]])"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.hierarchy.linkage(dm,metric='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "pdist() missing 1 required positional argument: 'X'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-128-177ee8ba8643>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspatial\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdistance\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpdist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: pdist() missing 1 required positional argument: 'X'"
     ]
    }
   ],
   "source": [
    "scipy.spatial.distance.pdist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
